### PR TITLE
90534

### PR DESCRIPTION
**Pull request by nks2109**
|Field|Value|
|-----|-----|
| Redmine Ticket Number | [90534](https://devops.dchbx.org/redmine/issues/90534) |
| App-Dev Road Map # | n/a |
| Start Date | n/a |
| Due Date | n/a  |
| App-Dev Priority | n/a |
| Development Story Points | n/a |
| Peer Review Story Points | n/a |
| Ticket Author | Sheheryar Tariq |

### CODE REVIEW

File: lib/tasks/plan_validation_reports.rake [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          carrier_name = carrier.name
+          carrier_id = carrier.plans.first.try(:hios_plan_id)[0..4]
+          plans = carrier.plans.where(year: active_year)
+          premium_tables = plans.map(&:premium_tables).flatten.select do |prem_table|
+            start_date = prem_table.rate_start_date.to_date
+            end_date = prem_table.rate_end_date.to_date
+            (start_date..end_date).cover?(active_date)
+          end
+          (0..120).each do |age|
+            age_cost = premium_tables.select{|a|a.age == age}.flatten.map(&:amount).sum
```
Comment (saimekala07) [inline lib/tasks/plan_validation_reports.rake]: @nks2109 Do we need flatten here..

File: lib/tasks/plan_validation_reports.rake [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    active_date = args[:active_date].to_date
+    active_year = active_date.year
+    CSV.open("#{Rails.root}/plan_validation_report2_#{active_year}.csv", "w", force_quotes: true) do |csv|
+      csv << %w[PlanYearId CarrierId CarrierName Age Sum]
+      Carrier.all.each do |carrier|
+        begin
+          next if carrier.plans.count < 1
+          carrier_name = carrier.name
+          carrier_id = carrier.plans.first.try(:hios_plan_id)[0..4]
+          plans = carrier.plans.where(year: active_year)
```
Comment (saimekala07) [inline lib/tasks/plan_validation_reports.rake]: @nks2109 Check plans exist for the carrier for a given year.

next unless plans.present?

Comment (nks2109) [reply lib/tasks/plan_validation_reports.rake]: sure

File: lib/tasks/plan_validation_reports.rake [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    puts "Reports generation started for Report1" unless Rails.env.test?
+    Rake::Task['plan_validation:report1'].invoke(args[:plan_year])
+    puts "Reports generation started for Report2" unless Rails.env.test?
+    Rake::Task['plan_validation:report2'].invoke(args[:active_date])
+  end
+
+  #To run first report: RAILS_ENV=production bundle exec rake plan_validation:report1["2020-01-01"]
+  desc "Details about Plan Count by Carrier, Coverage and Tier"
+  task :report1, [:plan_year] => :environment do |_task, args|
+    active_date = args[:active_date].to_date
```
Comment (saimekala07) [inline lib/tasks/plan_validation_reports.rake]: env args passed "plan_year", checking for "active_date"..fix this.

**Review by TreyE — CHANGES_REQUESTED** [review]
Pull this into a class or method, instead of hiding all the logic in a rake task.

I'd also like to see specs, as I'm not seeing any.

**Review by TreyE — APPROVED** [review]

**Pull request merged**